### ISSUE TITLE
p5 doesn't draw opaque strokes as expected (RGB(255,255,255,1) not working properly?)

### ISSUE THREAD

**Issue by KevinGutowski** [processing/p5.js#842]
Not sure if its just how p5 is interpreting my code or what but I can't seem to get it to draw strokes properly in colorMode(RGB,255,255,255,1). Here is the code that is giving me issues:

```javascript
function setup() {
	createCanvas(windowWidth, windowHeight);
	colorMode(RGB,255,255,255,1);
}

function draw() {
	background(255);
	noFill();
	strokeWeight(4);
	stroke(255,0,10,.48);
	ellipse(width/2,height/2,40,40);
	ellipse(width/2 + 10, height/2 + 10, 30,30);	
}
```

I essentially just draw two overlapping circles but what I get isn't right at all. It seems to fade away even though I am redrawing them??? Seems like a bug because I can achieve this effect just fine with similar code in Processing. Pretty bizarre!

Seems like when I switch over to fill() rather than stroke() I don't get any shapes to draw on screen. When I use colorMode(RGB,255,255,255) without the alpha it seems to work fine. Just seems like the alpha channel is messing some things up. 

Processing
```
void setup() {
  size(500,500);
  background(255);
  colorMode(RGB,255,255,255,1);
}

void draw() {
  background(255);
  noFill();
  strokeWeight(4);
  stroke(255,0,10,.3);
  ellipse(width/2,height/2,40,40);
  ellipse(width/2 + 10,height/2 + 10,30,30);
}
```

**Comment by maxkolyanov** [comment]
p5.js Color object handles alpha on the scale from 1 to 255.
The problem with this code is value you provide gets divided by 255 by this line of code
`var alpha = typeof a[3] !== 'undefined' ? a[3] / 255 : 1;` 
and comes out as very small ~0.0003, which canvas then renders as transparent. 

I would happily write a handler for floating point alpha channel, if community/contributors are okay with this idea. 
Personally I see this as something very viable, since setting alpha by providing a float is somewhat a standard.

**Comment by sarahgp** [comment]
You can change the alpha scale (or any color scale) by declaring it in your `colorMode()` call. So in theory, calling 

```
colorMode(255, 255, 255, 1)
```

should in fact result in alpha being calculated between 0 and 1. (I updated this in #728 ).

I can take a look sometime this week or if someone has time sooner and wants to jump on it, please feel free.

**Comment by maxkolyanov** [comment]
Whoops, I see code responsible for this now! 
Sorry for misinformation, still trying to grasp the code and architecture
Would try and look into actual cause.

**Comment by sarahgp** [comment]
I am not sure what you mean by "ignores". It shouldn't. When the color is set, it should normalize it against the max.

As far as the ability to put in ostensibly-nonsensical scales, that is functioning as intended. There is no reason to limit people.

**Comment by maxkolyanov** [comment]
@sarahgp 
```
p5.prototype.colorMode = function() {
  if (arguments[0] === constants.RGB ||
      arguments[0] === constants.HSB ||
      arguments[0] === constants.HSL) {
    this._colorMode = arguments[0];

    var maxArr = this._colorMaxes[this._colorMode];

    if (arguments.length === 2) {
      maxArr[0] = arguments[1];
      maxArr[1] = arguments[1];
      maxArr[2] = arguments[1];
      maxArr[3] = arguments[1];
    } else if (arguments.length > 2) {
      maxArr[0] = arguments[1];
      maxArr[1] = arguments[2];
      maxArr[2] = arguments[3];
    }
    if (arguments.length === 5) {
      maxArr[3] = arguments[4];
    }
  }
  return this;
};
```
What I mean is that maxArr never goes anywhere, it is just defined inside a function, filled with arguments and then left in the memory to sit all alone.
I could write `colorMode(RGB,'asd','asd','asd)` or whatever, it won't affect anything

**Comment by sarahgp** [comment]
So then I suspect it is being applied elsewhere -- or else for some reason it is not being applied when just an RGB color is declared and that is a bug that we missed in our test cases. I can help you figure out which later this week or you can read through the rest of the color files and find out why the tests seem to pass but this is failing if you are feeling impatient. :smiley:

**Comment by maxkolyanov** [comment]
@sarahgp 
What I did is: I searched code for usage of `_colorMaxes` array and it's almost never even used nor modified anywhere, let alone in normalising colors.
The tests are passing, because there is no coverage for this case. 
Setting anything with `colorMode()`, won't affect instantiation of Color object in any way, as far as I can see.
I also can be absolutely wrong, but if this is really the case, then I would be glad to work on implementing the stuff.

**Comment by sarahgp** [comment]
Okay, so I see what is going on; there are a few changes we need to make. 

The main problem with this bug is that when an RGB color is used with a set of maxes other than `[255, 255, 255, 255]`, it is never getting normalized. Normalization is applied in the conversion functions in `color_utils.js` but this doesn't trigger the conversion function. It probably makes sense to add that normalization to the `p5.Color._getFormattedColor` method in `p5.Color.js`. Or perhaps writing a short standalone normalization function and then adding to the `isRGB` case in the `p5.Color` constructor itself.

We missed this because the test of "new p5.Color in RGB-illogical mode" in the `p5.Color` unit tests involves getting out the hue, saturation and brightness ... which trigger normalization. This end of the file is kinda a mess (mea maxima culpa!), with an empty test and a set of base 1 tests commented out pending some unspecified discussion. So while we are fixing the bug, we should update the tests.

As for the `colorMode()` function pasted above, it is actually _setting_ the `_colorMaxes` values. `maxArr` is being used as a shorthand reference for ` this._colorMaxes[this._colorMode]` — since objects are passed by reference, it is reaching back and mutating the array that is then used in `p5.Color` functions, particularly the conversions. We should probably add a comment here explaining that since it is definitely ambiguous.

So we should probably put together a PR covering all these issues. I mostly likely won't have a chance till this weekend or next week, so if you want to jump in, @maxkolyanov, I think that's what needs to be done.

**Comment by sarahgp** [comment]
Another thing I noticed in looking at `_getFormattedColor`, we are still [hardcoding 255](https://github.com/processing/p5.js/blob/master/src/color/p5.Color.js#L485) to set unmatched values to white, but it should be moved to dynamic maxes, otherwise it could default to something not-white depending on the scale.

**Comment by maxkolyanov** [comment]
Cool! 
I'll try to put all these changes into a PR, when I have more time this week to look thoroughly around the code.

**Comment by sarahgp** [comment]
I didn't realized you added to that previous post as well. The `toString()` method you linked to has its own discussion going on in #723. I would not read anything into expected behavior from that particular method: it should just be output. 

But yeah, anywhere where we are using a hardcoded 255 is likely a bug waiting to happen. When I added HSLA, I changed all the hardcoded places I came across but I definitely missed things!

**Comment by sarahgp** [comment]
And thanks!

**Comment by KevinGutowski** [comment]
Thanks for the quick responses everyone!

**Comment by lmccart** [comment]
closed with #857
